src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -379,8 +370,45 @@ internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeT
                     enumString = reader.GetString();
                 }
 
+                if (enumString == null)
+                {
+                    ThrowHelper.ThrowJsonException();
+                }
+
                 Debug.Assert(_enumValueCache != null, "Enum value cache should be instantiated if a naming policy is specified.");
-                if (enumString == null || !_enumValueCache.TryGetValue(enumString, out value))
+
+                if (!enumString.Contains(ValueSeparator))
+                {
+                    success = _enumValueCache.TryGetValue(enumString, out value);
+                }
+                else
+                {
+                    // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
+                    string[] enumValues = enumString.Split(
+#if BUILDING_INBOX_LIBRARY
```

Comment:
FYI https://github.com/dotnet/runtime/pull/73443 will remove `BUILDING_INBOX_LIBRARY`. `NETCOREAPP` is equivalent and will avoid any merge conflicts.
```suggestion
#if NETCOREAPP
```